source/opt/simplification_pass.cpp

```diff
@@ -44,13 +60,15 @@ bool SimplificationPass::SimplifyFunction(Function* function) {
   std::unordered_set<Instruction*> process_phis;
   std::unordered_set<Instruction*> inst_to_kill;
   std::unordered_set<Instruction*> in_work_list;
+  std::unordered_set<Instruction*> inst_seen;
```

Comment:
OK. In the meantime I will remove the "efficiency" checks to see if that looks better to you.

Reply:
> Another way of doing it would be to note that any new instruction will be a block preceding the current instruction.

That doesn't sound right to me. Generally the folding rules just update instructions in place.

For the "factor" rule, I use a Builder to insert the new instruction, and then update the original instruction in place. I don't think any new blocks are created.

Reply:
I did not mean a new basic block.  I just meant that all of the new instructions would be contiguous.  Your code will be fine for now.